### PR TITLE
Drop a PVC finalizer to delete pods

### PR DESCRIPTION
**Pull request by peng225**
issue: topolvm/topolvm#526
Signed-off-by: Shinya Hayashi <shinya-hayashi@cybozu.co.jp>

### CODE REVIEW

**Review by toshipp — COMMENTED** [review]
https://github.com/topolvm/topolvm/pull/536/files#diff-6b58a5c82d93f021da0d7a434a9c487717fd238387e83248e58ab970735a64dbR19
This line is no longer needed.

File: controllers/persistentvolumeclaim_controller.go [hunk 1]
```diff
@@ -127,3 +61,18 @@ func (r *PersistentVolumeClaimReconciler) SetupWithManager(mgr ctrl.Manager) err
 		For(&corev1.PersistentVolumeClaim{}).
 		Complete(r)
 }
+
+func removeTopoLVMFinalizer(pvc *corev1.PersistentVolumeClaim) ([]string, bool) {
+	removed := false
+	// Due to bug #310, we need to de-dup TopoLVM finalizers.
```
Comment (toshipp) [inline controllers/persistentvolumeclaim_controller.go:67]: This comment does not describe what the code does. please update or remove it.

File: controllers/persistentvolumeclaim_controller_test.go [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+		reconciler := PersistentVolumeClaimReconciler{
+			Client: k8sClient,
+		}
+		err = reconciler.SetupWithManager(mgr)
+		Expect(err).NotTo(HaveOccurred())
+
+		ctx, cancel := context.WithCancel(ctx)
+		stopFunc = cancel
+		go func() {
+			err := mgr.Start(ctx)
```
Comment (toshipp) [inline controllers/persistentvolumeclaim_controller_test.go:43]: To stop it properly, why don't you use a chan to signal its stop?

**Review by toshipp — CHANGES_REQUESTED** [review]

**Comment by peng225** [comment]
@toshipp 
> https://github.com/topolvm/topolvm/pull/536/files#diff-6b58a5c82d93f021da0d7a434a9c487717fd238387e83248e58ab970735a64dbR19
This line is no longer needed.

From the link you shared, I could not understand to which line you metioned. Could you check the link, or tell me which line of which file I should remove?

File: controllers/persistentvolumeclaim_controller.go (hunk 1 above)
Comment (peng225) [reply controllers/persistentvolumeclaim_controller.go]: I revised the comment.

**Comment by toshipp** [comment]
It seems that a link to a folded line will be invalid...
```
// PersistentVolumeClaimReconciler reconciles a PersistentVolumeClaim object
type PersistentVolumeClaimReconciler struct {
	client.Client
	APIReader client.Reader // <- IIUC, this is unnecessary now
}
```

File: controllers/persistentvolumeclaim_controller.go [hunk 3]
```diff
@@ -37,83 +36,18 @@ func (r *PersistentVolumeClaimReconciler) Reconcile(ctx context.Context, req ctr
...
-		return ctrl.Result{}, err
-	}
-	for _, pod := range pods {
-		err := r.Delete(ctx, &pod, client.GracePeriodSeconds(1))
-		if err != nil {
-			log.Error(err, "unable to delete Pod", "name", pod.Name, "namespace", pod.Namespace)
+	var removed bool
+	pvc.Finalizers, removed = removeTopoLVMFinalizer(pvc)
+	if removed {
+		if err := r.Update(ctx, pvc); err != nil {
```
Comment (daichimukai) [inline controllers/persistentvolumeclaim_controller.go:42]: Why not do these things in `removeTopoLVMFinalizer()`?

File: controllers/persistentvolumeclaim_controller.go [hunk 4]
```diff
@@ -127,3 +61,19 @@ func (r *PersistentVolumeClaimReconciler) SetupWithManager(mgr ctrl.Manager) err
...
 }
+
+func removeTopoLVMFinalizer(pvc *corev1.PersistentVolumeClaim) ([]string, bool) {
+	removed := false
+	// Due to the bug #310, multiple TopoLVM finalizers can exist in `pvc.Finalizers`.
+	// So we need to delete all of them.
+	for i := 0; i < len(pvc.Finalizers); {
+		if pvc.Finalizers[i] == topolvm.PVCFinalizer {
+			copy(pvc.Finalizers[i:], pvc.Finalizers[i+1:])
+			pvc.Finalizers = pvc.Finalizers[:len(pvc.Finalizers)-1]
```
Comment (daichimukai) [inline controllers/persistentvolumeclaim_controller.go:72]: nits: might be simplified like

```suggestion
			pvc.Finalizers = append(pvc.Finalizers[:i], pvc.Finalizers[i+1:]...)
```

File: controllers/persistentvolumeclaim_controller_test.go [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+	"github.com/topolvm/topolvm"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+var _ = Describe("PersistentVolumeClaimController controller", func() {
+
```
Comment (daichimukai) [inline controllers/persistentvolumeclaim_controller_test.go:19]: nits: remove this empty line

File: controllers/persistentvolumeclaim_controller_test.go [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+		}).Should(Succeed())
+	})
+})
+
+func newPVCWithFinalizer() *corev1.PersistentVolumeClaim {
+	return &corev1.PersistentVolumeClaim{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:       "sample",
+			Namespace:  "test",
+			Finalizers: []string{topolvm.PVCFinalizer, "topolvm.io/finalizer1", topolvm.PVCFinalizer, topolvm.PVCFinalizer, "topolvm.io/finalizer3", topolvm.PVCFinalizer},
```
Comment (daichimukai) [inline controllers/persistentvolumeclaim_controller_test.go:84]: Why 3?

File: controllers/persistentvolumeclaim_controller_test.go (hunk 2 above)
Comment (peng225) [reply controllers/persistentvolumeclaim_controller_test.go]: Though I tried to fix it, I'm not confident with my understainding.
Are the changes in the following commit what you intended?

7d4eb1bd6721551f46d11cb1fe36df12578c188f

**Comment by peng225** [comment]
@toshipp 
> It seems that a link to a folded line will be invalid...

Thanks. I removed the line and a related code in `pkg/topolvm-controller/cmd/run.go`

File: controllers/persistentvolumeclaim_controller_test.go (hunk 6 above)
Comment (peng225) [reply controllers/persistentvolumeclaim_controller_test.go]: I changed it 2.

File: controllers/persistentvolumeclaim_controller_test.go (hunk 5 above)
Comment (peng225) [reply controllers/persistentvolumeclaim_controller_test.go]: I removed it.

File: controllers/persistentvolumeclaim_controller.go (hunk 4 above)
Comment (peng225) [reply controllers/persistentvolumeclaim_controller.go]: Thanks. I fixed it as you proposed.

File: controllers/persistentvolumeclaim_controller.go (hunk 3 above)
Comment (peng225) [reply controllers/persistentvolumeclaim_controller.go]: Thanks. I moved these lines to `removeTopoLVMFinalizer`.

File: controllers/persistentvolumeclaim_controller_test.go (hunk 2 above)
Comment (toshipp) [reply controllers/persistentvolumeclaim_controller_test.go]: I mean use a chan to transfer err value.
For example:

```
var errCh chan err
BeforeEach(func() {
  ...
  errCh <- mgr.Start(ctx)
})

AfterEach(func() {
  stopFunc()
  <- errCh
   ...
})
```
WaitGroup is also good.

BTW, why do you use `for` to start mgr?

Comment (peng225) [reply controllers/persistentvolumeclaim_controller_test.go]: Thank you for your explanation. I got it.
I modified the code in a similar way to what you showed.

**Review by toshipp — DISMISSED** [review]

**Review by daichimukai — APPROVED** [review]

**Comment by daichimukai** [comment]
@peng225 Please rebase commits, or merge this PR if you are happy with as-is.

**Comment by peng225** [comment]
@daichimukai @toshipp 
> Please rebase commits, or merge this PR if you are happy with as-is.

I squashed commits into two commits. (File contents are not changed.)
Please check it.

**Review by toshipp — APPROVED** [review]

**Review by daichimukai — APPROVED** [review]

**Pull request merged**